crates/ruma-federation-api/src/authenticated_media.rs

```diff
@@ -0,0 +1,291 @@
+//! Authenticated endpoints for the content repository, according to [MSC3916].
+//!
+//! [MSC3916]: https://github.com/matrix-org/matrix-spec-proposals/pull/3916
+
+use serde::{Deserialize, Serialize};
+
+pub mod get_content;
+pub mod get_content_thumbnail;
+
+/// The metadata of a file from the content repository.
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct ContentMetadata {}
+
+impl ContentMetadata {
+    /// Creates a new empty `ContentMetadata`.
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// A file from the content repository or the location where it can be found.
+#[derive(Debug, Clone)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum FileOrLocation {
+    /// The content of the file.
+    File(Content),
+
+    /// The file is at the given URL.
+    Location(String),
+}
+
+/// The content of a file from the content repository.
+#[derive(Debug, Clone)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct Content {
+    /// The content of the file as bytes.
+    pub file: Vec<u8>,
+
+    /// The content type of the file that was previously uploaded.
+    pub content_type: String,
+
+    /// The value of the `Content-Disposition` HTTP header, possibly containing the name of the
+    /// file that was previously uploaded.
+    ///
+    /// See [MDN] for the syntax.
+    ///
+    /// [MDN]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#Syntax
+    pub content_disposition: String,
+}
+
+impl Content {
+    /// Creates a new `Content` with the given bytes, content type and content disposition.
+    pub fn new(file: Vec<u8>, content_type: String, content_disposition: String) -> Self {
+        Self { file, content_type, content_disposition }
+    }
+}
+
+/// Serialize the given metadata and content into a `http::Response` `multipart/mixed` body.
+///
+/// Returns a tuple containing the boundary used
+#[cfg(feature = "client")]
+fn try_into_multipart_mixed_response<T: Default + bytes::BufMut>(
+    metadata: &ContentMetadata,
+    content: FileOrLocation,
+) -> Result<http::Response<T>, ruma_common::api::error::IntoHttpError> {
+    use mail_builder::{headers::text::Text, mime::MimePart};
+
+    let metadata_raw = serde_json::to_vec(&metadata)?;
```

Comment:
Is there a reason we are passing `&&ContentMetadata` into `to_vec`, instead of just `&ContentMetadata`?

```suggestion
    metadata: &ContentMetadata,
    content: FileOrLocation,
) -> Result<http::Response<T>, ruma_common::api::error::IntoHttpError> {
    use mail_builder::{headers::text::Text, mime::MimePart};

    let metadata_raw = serde_json::to_vec(metadata)?;
```

Reply:
No, just something remaining from when I started implementing it. I thought clippy would catch something like that.

Reply:
Fixed.